optaplanner-quarkus-integration/optaplanner-quarkus/runtime/src/main/java/org/optaplanner/quarkus/nativeimage/Substitute_ConfigUtils.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.quarkus.nativeimage;
+
+import java.lang.reflect.Modifier;
+
+import org.optaplanner.quarkus.gizmo.OptaPlannerGizmoBeanFactory;
+
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+
+@TargetClass(className = "org.optaplanner.core.config.util.ConfigUtils")
+public final class Substitute_ConfigUtils {
+
+    @Substitute
+    public static <T> T newInstance(Object bean, String propertyName, Class<T> clazz) {
+        T out = OptaPlannerGizmoBeanFactory.INSTANCE.getReference().newInstance(clazz);
+        if (out != null) {
+            return out;
+        } else {
+            throw new IllegalArgumentException("The " + bean.getClass().getSimpleName() + "'s " + propertyName + " ("
+                    + clazz.getName() + ") does not have a public no-arg constructor"
+                    // Inner classes include local, anonymous and non-static member classes
+                    + ((clazz.isLocalClass() || clazz.isAnonymousClass() || clazz.isMemberClass())
+                            && !Modifier.isStatic(clazz.getModifiers()) ? " because it is an inner class." : "."));
```

Comment:
What is the user supposed to do with an exception like this? What could the user do?

Reply:
This was to duplication the error behavior of ConfigUtils: https://github.com/kiegroup/optaplanner/blob/fb88c87e1799505b902f1f144235277a7a32f349/optaplanner-core/src/main/java/org/optaplanner/core/config/util/ConfigUtils.java#L58-L65 . Should never happen normally (unless ConfigUtils is used to create a class that wasn't used in an annotation/we forget to add the annotation to the detector.

Reply:
We usually do one of the two things with exceptions:

- If it is unrecoverable, we usually add the prefix "Impossible state: " to tell the user that this is not their fault, but our bug.
- If it could, however, be their fault, we add a new line to the message, saying "Maybe... " and pointing them to what they could try to fix. See other uses of "maybe" in the codebase.

Reply:
Changed the message to "Impossible State: could not find the " + bean.getClass().getSimpleName() +
                    "'s " + propertyName + " (" + clazz.getName() + ") generated Gizmo supplier."